### DISCUSSION TITLE
SSL Certs behind Nginx Proxy Manager (Separate Servers)

### DISCUSSION THREAD

**Discussion by Xisdibik** [Ylianst/MeshCentral#6300] [Issues]
Hello,

I have Meshcentral on a VM (debian) on my proxmox server that when originally setup behind nginx but not setup to hand off SSL to it, allows access to the server page but agents won't connect.  I then followed the steps in the userguide to get nginx and the server speaking to each other and either ended with 502 Gateway error (before) or ERR_SSL_UNRECOGNIZED_NAME_ALERT (current).  I have looked through a lot of similar issues and tried some of their fixes but none of them seem to help or are not exactly what I am facing.

Am I supposed to be downloading the cert from Nginx and placing it somewhere (replace the webserver/agent crt files?) or something like that?

The only thing I didn't do with the nginx instructions (guide + video) was do the 4430 and 800 ports because they are not on the same machine.

my config.json looks like this (mesh.mydomain.com is not whats actually there it is the real domain name)

```
{
  "$schema": "https://raw.githubusercontent.com/Ylianst/MeshCentral/master/meshcentral-config-schema.json",
  "__comment1__": "This is a simple configuration file, all values and sections that start with underscore (_) are ignored. Edit a section and remove the _ in front of the name. Refer to the user's guide for details.",
  "__comment2__": "See node_modules/meshcentral/sample-config-advanced.json for a more advanced example.",
  "settings": {
    "MongoDb": "mongodb://127.0.0.1:27017/meshcentral",
    "cert": "mesh.mydomain.com",
    "WANonly": true,
    "_LANonly": true,
    "_sessionKey": "MyReallySecretPassword1",
    "_port": 443,
    "_aliasPort": 443,
    "_redirPort": 80,
    "_redirAliasPort": 80,
    "agentPong": 300,
    "TlsOffload": "192.168.85.30",
    "trustedProxy": "192.168.85.30"
  },
  "domains": {
    "": {
      "_title": "MyServer",
      "_title2": "Servername",
      "_minify": true,
      "_newAccounts": true,
      "_userNameIsEmail": true,
      "certUrl": "https://192.168.85.30/"
    }
  },
  "_letsencrypt": {
    "__comment__": "Requires NodeJS 8.x or better, Go to https://letsdebug.net/ first before trying Let's Encrypt.",
    "email": "myemail@mydomain.com",
    "names": "myserver.mydomain.com",
    "skipChallengeVerification": true,
    "production": false
  }
}
```

Anyone have any idea where I am going wrong?

**Comment by si458** [comment]
so from the looks your config is all correct apart from the certurl should be the DNS name you use to access meshcentral,
and make sure it can be resolved from within the meshcentral VM! `curl -v https://mymeshserver.com`

then you should point your nginx to http://MESHCENTRALVMIP:443, make sure its HTTP and NOT HTTPS!

**Reply by Xisdibik** [reply]
running your curl command seems to connect.  It does gives this for the last 2 lines:

`* OpenSSL/3.0.13: error:0A00010B:SSL routines::wrong version number
* Closing connection 0
curl: (35) OpenSSL/3.0.13: error:0A00010B:SSL routines::wrong version number`

Not sure if that helps with anything.

I changed the certurl to the DNS name i use to access meshcentral, and then updated nginx  with http the the domain and then port 443.  I unfortunately still get the same issue.  

One thing i forgot to mention above is that I am using Nginx Proxy Manager and not just Nginx.  I wasn't sure if I should put the nginx config that was supplied in some specific location on the server or put it in the "advanced" tab on the nginx proxy manager webpage for the specific mesh proxy.  (its in the advanced tab, should it be elsewhere?)

**Reply by si458** [reply]
What's the output when u start meshcentral in the vm?
It should show what ports it running etc

**Reply by Xisdibik** [reply]
`MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, WAN mode.
Failed to load web certificate at: "http://mesh.mydomain.com/", host: "mesh.mydomain.com"
MeshCentral Intel(R) AMT server running on mesh.mydomain.com:4433.
MeshCentral HTTP server running on port 443.`

**Reply by si458** [reply]
`Failed to load web certificate at: "http://mesh.mydomain.com"`

You should set the `certurl` as `https://mesh.mydomain.com`
Then restart meshcentral 

Can u access ur meshcentral vm directly `http://meshcentralipaddress:443`

**Reply by Xisdibik** [reply]
changed certurl to `https://mesh.mydomain.com`.  restarted meshcentral.  same issue.

Yes I can access meshcentral directly by going to `http://meshcentralip:443`

**Reply by Xisdibik** [reply]
Also if this helps.  I have a cloudflare Let's encrypt cert that is *.mydomain.com   Do i need to make a separate SSL cert for mesh.mydomain.com or should that wildcard cert work in this situation? (should have said this earlier sorry)

**Reply by si458** [reply]
No, the certificate should be fine.
You just need to work out why meshcentral can't access ur reverse proxy now!
So if u do `cert -v https://mesh mydomain.com` on ur meshcentral server, u should get a reply of some form,
What's the whole output?

**Reply by Xisdibik** [reply]
sorry did you mean `curl` like before?  `cert` says no command.

the original full output for `curl` was this:

` Trying 127.0.1.1:443...
Connected to mesh.mydomain.com (127.0.1.1) port 443 (#0)
ALPN: offers h2,http/1.1
TLSv1.3 (OUT), TLS handshake, Client hello (1):
 CAfile: /etc/ssl/certs/ca-certificates.crt
*CApath: /etc/ssl/certs
OpenSSL/3.0.13: error:0A00010B:SSL routines::wrong version number
Closing connection 0
curl: (35) OpenSSL/3.0.13: error:0A00010B:SSL routines::wrong version number `

however when I ran it just now I got this:

` Trying 127.0.1.1:443...
connect to 127.0.1.1 port 443 failed: Connection refused
Failed to connect to mesh.mydomain.com port 443 after 0 ms: Couldn't connect to server
Closing connection 0
curl: (7) Failed to connect to mesh.mydomain.com port 443 after 0 ms: Couldn't connect to server `

not sure why it won't connect now.

Got code working, but had to remove some *'s.  so if you prefer it not in  "code" let me know.

**Reply by si458** [reply]
Ah, right, so your machine name is the same name as ur dns name, so it's trying to connect to itself! So that won't work!

You need to edit the `/etc/hosts` and set `reverseproxyip mesh.mydomain.com`

Then try the curl command again

**Reply by Xisdibik** [reply]
Thanks!

now it says this:

`*   Trying 127.0.1.1:443...
* connect to 127.0.1.1 port 443 failed: Connection refused
*   Trying 192.168.85.30:443...
* Connected to mesh.mydomain.com (192.168.85.30) port 443 (#0)
* ALPN: offers h2,http/1.1
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (IN), TLS alert, unrecognized name (624):
* OpenSSL/3.0.13: error:0A000458:SSL routines::tlsv1 unrecognized name
* Closing connection 0
curl: (35) OpenSSL/3.0.13: error:0A000458:SSL routines::tlsv1 unrecognized name`

Forget the last part, i know why.

**Reply by si458** [reply]
That should be OK!
Now restart meshcentral and check its output on startup again, is it any different now?

**Reply by Xisdibik** [reply]
Now it shows this:

`node ./node_modules/meshcentral
MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, WAN mode.
MeshCentral Intel(R) AMT server running on mesh.fixxix.com:4433.
Failed to load web certificate at: "https://mesh.mydomain.com", host: "mesh.mydomain.com"
MeshCentral HTTP server running on port 443.`

**Reply by si458** [reply]
OK so it's still failing

So one thing to check is the curl did say `connect to 127.0.1.1 port 443 failed: Connection refused` so check the `/etc/hosts` and make sure you only have a single entry for `mesh.mydomain.com`!

You might need to remove the 127.0.1.1 value if its there!

Then u need to access your website url on a different computer, don't worry it might still time out, but check its ssl certificate is valid!
Can u share screenshots of this?

**Reply by Xisdibik** [reply]
Removed the 127.0.0.1 entry. 
restarted mesh. 

`MeshCentral HTTP redirection server running on port 80.
MeshCentral v1.1.27, WAN mode.
MeshCentral Intel(R) AMT server running on mesh.fixxix.com:4433.
MeshCentral HTTP server running on port 443.
Failed to load web certificate at: "https://mesh.mydomain.com", host: "mesh.mydomain.com"
`

Tried connecting from my phone and got this (same as before)

![IMG_8883](https://github.com/user-attachments/assets/fa08d2b2-8628-4c26-b848-2915a06691e0)

**Reply by si458** [reply]
OK so the is something wrong with ur nginx now, and it's ssl!

So can u share ur nginx config/setup pages/ screenshots etc?

**Reply by Xisdibik** [reply]
![Nginx1](https://github.com/user-attachments/assets/f1d999d0-fe99-4ca2-b3af-3371e27658ca)

![Nginx2](https://github.com/user-attachments/assets/70ca761b-342c-431c-9e47-a60db00662e5)


The "Advanced Tab" is where i have this code from the user guide for meshcentral.

```
worker_processes 1;

events {
    worker_connections 1024;
}

http {
    # HTTP server. In this example, we use a wildcard as server name.
    server {
        listen 80;
        server_name mesh.mydomain.com;

        location / {
            proxy_pass http://192.168.85.33:80/;
            proxy_http_version 1.1;

            # Inform MeshCentral about the real host, port and protocol
            proxy_set_header X-Forwarded-Host $host:$server_port;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }
    }

    # HTTPS server. In this example, we use a wildcard as server name.
    server {
        listen 443 ssl;
        server_name mesh.mydomain.com;

        # MeshCentral uses long standing web socket connections, set longer timeouts.
        proxy_send_timeout 330s;
        proxy_read_timeout 330s;

        # We can use the MeshCentral generated certificate & key
        ssl_certificate webserver-cert-public.crt;
        ssl_certificate_key webserver-cert-private.key;
        ssl_session_cache shared:WEBSSL:10m;
        ssl_ciphers HIGH:!aNULL:!MD5;
        ssl_prefer_server_ciphers on;

        location / {
            proxy_pass http://192.168.85.33:443/;
            proxy_http_version 1.1;

            # Allows websockets over HTTPS.
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
            proxy_set_header Host $host;

            # Inform MeshCentral about the real host, port and protocol
            proxy_set_header X-Forwarded-Host $host:$server_port;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }
    }
}
```

**Reply by Xisdibik** [reply]
My guess is the issue lies in the advanced tab.  Perhaps I need to remove the section labeled "We can use the MeshCentral generated cert and key"  since we are using the one from Nginx?

**Reply by si458** [reply]
I don't think u need to add ANYTHING AT ALL in the advanced tab! So remove it all. Then try accessing ur Web url from ur phone/laptop again

**Reply by Xisdibik** [reply]
removed the nginx stuff and now i get "400 Bad Request" "The plain HHTP Request was sent to HTTPS port".  do i need to have nginx doing https instead of http again?

**Reply by si458** [reply]
Your nginx isn't setup correctly to redirect http to https, so ur trying to access it using http://mesh.mydomain.com
Try actually typing `https://mesh.mydomain.com` in the address bar on ur mobile/comp

**Reply by Xisdibik** [reply]
That's what I did.  `https://mesh.mydomain.com` and it gives the 400 Bad Request  the plain http request was sent to https port openresty

**Reply by si458** [reply]
Daft one, restart meshcentral now, does the output from it now change?
What about accessing the url after restarting meshcentral?

**Reply by Xisdibik** [reply]
rebooted meshcentral, same response both from my desktop (the one I am typing here but not the same one the VM is on) and my phone.

Also rebooted the whole VM, and still same result.

**Reply by si458** [reply]
What about turning off the force ssl option in the ssl tab of nginx proxy manager?

**Reply by Xisdibik** [reply]
Turning force SSL off gives the same result

**Reply by si458** [reply]
Are u using cloudflare and it's tunnels?
U might need to go into the tunnel settings in cloudflare, then into tls settings then turn on no tls verify

**Reply by si458** [reply]
One more thing u can also try is remove the tlsoffload option from config.json in meshcentral, restart meshcentral, then change nginx to https and enable force ssl

**Reply by Xisdibik** [reply]
Will try the stuff you just mentioned but as for cloudflare.  Yes I am using them for the Let's Encrypt SSL cert.  I don't think I am using its tunnel service though.  and the domain is not proxied through them (DNS only) for the A record on my domain.

**Reply by Xisdibik** [reply]
commenting out tlsoffload in the config.json and rebooting meshcentral and then updating nginx to https and enabling force ssl gives me now 502 gateway error when trying to visit the mesh page

**Reply by si458** [reply]
> Will try the stuff you just mentioned but as for cloudflare.  Yes I am using them for the Let's Encrypt SSL cert.  I don't think I am using its tunnel service though.  and the domain is not proxied through them (DNS only) for the A record on my domain.

Right that's OK! I wasn't sure if u was using there tunnels product to avoid opening ur firewall!

But if ur firewalls open and u are port forwarding to ur reverse proxy then that's OK too!

**Reply by Xisdibik** [reply]
Yep, I am just port forwarding the ports for nginx on my firewall.

**Reply by si458** [reply]
I will have to test tomorrow if I get chance (off bed)
Ive never used nginx proxy manager before so u might need something specific setting somewhere!

**Reply by Xisdibik** [reply]
That was my thought and I searched high and low for someone using proxy manager to give me a hint but was unsuccessful.  Have a good night (its just after noon here), I'll check back here later/tomorrow to see if you had any post sleep miracles.

**Reply by Xisdibik** [reply]
I was able to get it to bring up mesh central with the settings you provided (with minor adjustment, I will have to look through what I did since its been a few days of busy work.  However the meshagents still don't show up when connecting (which was the original issue).  So something I think does need to go in the nginx advanced tab (or location) to tell it to pass the SSL info to mesh central so the agents SSL certs match things.

Hopefully I'll have some free time soon to take another stab at it.
